sdk/core/azure-core-http-okhttp/src/main/java/com/azure/core/http/okhttp/implementation/OkHttpFluxRequestBody.java

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.http.okhttp.implementation;
+
+import com.azure.core.implementation.util.BinaryDataContent;
+import com.azure.core.util.logging.ClientLogger;
+import okhttp3.MediaType;
+import okhttp3.OkHttpClient;
+import okio.BufferedSink;
+import reactor.core.Exceptions;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Schedulers;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.time.Duration;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * An {@link okhttp3.RequestBody} subtype that sends {@link BinaryDataContent}
+ * as {@link Flux} of {@link ByteBuffer} in an unbuffered manner.
+ * This class accepts any {@link BinaryDataContent} as catch-all for backwards compatibility
+ * but ideally should be used only with reactive payloads.
+ */
+public class OkHttpFluxRequestBody extends OkHttpStreamableRequestBody<BinaryDataContent> {
+
+    private static final ClientLogger LOGGER = new ClientLogger(OkHttpFluxRequestBody.class);
+
+    private final AtomicBoolean bodySent = new AtomicBoolean(false);
+    private final OkHttpClient okHttpClient;
+
+    public OkHttpFluxRequestBody(
+        BinaryDataContent content, long effectiveContentLength, MediaType mediaType, OkHttpClient okHttpClient) {
+        super(content, effectiveContentLength, mediaType);
+        this.okHttpClient = Objects.requireNonNull(okHttpClient, "'okHttpClient' cannot be null.");
+    }
+
+    @Override
+    public void writeTo(BufferedSink bufferedSink) throws IOException {
+        if (bodySent.compareAndSet(false, true)) {
+            Mono<BufferedSink> requestSendMono = content.toFluxByteBuffer()
+                .publishOn(Schedulers.boundedElastic())
```

Comment:
I guess this thread switch to Bounded-Elastic-Thread is justified? Without this, Flux that is not backed by the user thread will use OkHttp-Thread (running `writeTo`) to emit, hence, using the same OkHttp-Thread to notify IOException downstream. By using Bounded-Elastic-Thread, the OkHttp-Thread (from the bounded IO thread pool?) is not obligated to run the entire error handling async chain but is immediately available for handling another request.

Not sure if there is a different reason, does it worth adding a code comment for future reference?

Reply:
@anuchandy The `bufferedSink` write is potentially blocking operation (got flagged by IDE) i.e. it will block when the sink reaches internal buffer size and emits data to the socket.

I was playing with this and depending on how incoming Flux is created the write may land on reactor's non-blocking thread. Hence being defensive here.

Ideally we'd detect that we're on non-blocking thread and push writes on bounded elastic only if that happens. I haven't found a good way to do that - do you know if we could do something like this?

Reply:
I see; thanks! I didn't realize that is a block blocking call hence the offload to Bounded Elastic Thread.

Reactor seems to expose this contract [NonBlocking](https://github.com/reactor/reactor-core/blob/v3.3.0.RELEASE/reactor-core/src/main/java/reactor/core/scheduler/NonBlocking.java#L28) which Reactor-Non-blocking threads implement. 

For example, this [Public API](https://github.com/reactor/reactor-core/blob/df31c608b44a0cd14175f9248dbd3fc1b0dee7fe/reactor-core/src/main/java/reactor/core/scheduler/Schedulers.java#L593) is checking if the Thread implements this contract.

Reply:
Would something like this work?

The downside that I can see is that flatMapSequential would involve internal queue. "reduce" seems to be less involved allocation wise.

```
content.toFluxByteBuffer()
                .flatMapSequential(buffer -> {
                    Mono<Void> writer = Mono.fromRunnable(() -> {
                        try {
                            while (buffer.hasRemaining()) {
                                bufferedSink.write(buffer);
                            }
                            return bufferedSink;
                        } catch (IOException e) {
                            throw Exceptions.propagate(e);
                        }
                    });
                    if (Schedulers.isInNonBlockingThread()) {
                        return Mono.just(1)
                            .publishOn(Schedulers.boundedElastic())
                            .then(writer);
                    } else {
                        return writer;
                    }
                });
 ```

Reply:
That seems the right operator combination to achieve the goal. 

Agree with the downside you pointed out; I'm not sure if the allocation_cost(new operator combination) > allocation_cost(publishOn) + threading

There seems to be an overload of `flatMapSequential` where we can set concurrency, prefetch to 0, that uses zero_size_Queue, that could help save alloc when reactor does fusion? (if we decide to go non-blocking-thread-check route).

Reply:
I'll run benchmark.

Reply:
Prefetch 0 is illegal but "ONE_SUPPLIER" queue ain't bad.

This is what I'm going to try:
```
            Mono<Void> requestSendMono = content.toFluxByteBuffer()
                .flatMapSequential(buffer -> {
                    if (Schedulers.isInNonBlockingThread()) {
                        return Mono.just(buffer)
                            .publishOn(Schedulers.boundedElastic())
                            .map(b -> writeBuffer(bufferedSink, b))
                            .then();
                    } else {
                        writeBuffer(bufferedSink, buffer);
                        return Mono.empty();
                    }
                }, 1, 1)
                .then();
```

Reply:
Nice! new code is more readable compared to the previous. 

> I'll run benchmark.

It will be interesting to understand. 

Just sharing some thoughts -

The `publishOn` operator obtains a Worker from the WorkersPool at the subscription time and uses the same Worker to publish until unsubscribed.

While with the original code, this acquiring happens once for the entire request-body; with the thread check logic, the publishOn reaches out to the WorkersPool for each chunk of the request-body.  (that's my understanding by scanning the code).

The frequent "acquiring workers" may not be a problem allocation wise because, over the long run, the global WorkersPool will cache the Workers (+ associated objects (e.g., work queue)). But probably more contention on acquiring paths when other parts of the application also race for the Workers.

Reply:
it a tradeoff between:
- not checking and potentially blocking reactor non-blocking threads (bad for the whole application)
- always hopping (bad for majority of requests)
- hopping per buffer (bad for some requests)

the last one seems to have smallest blast radius. so we should stick to it unless we find a solution to conditionally check the whole flux - but that's probably impossible (given that upstream can have as crazy logic as we have, i.e. each flux element might be in theory published on different thread...).

Reply:
Yes, I agree with you on this. Also, the thread(s) blocked/busy-looping for Worker acquisition is more likely to be unblocked faster. Because of those running requests (taking time to produce per chunk or having a large number of chunks), releasing workers after each chunk gives others chances to acquire it. Of course, it's a trade-off as you listed, and we should stick to it.